src/ui/reducers/hitCounts.ts

```diff
@@ -0,0 +1,180 @@
+  hitCounts?: HitCount[];
+  id: string;
+  status: LoadingState;
+}
+
+export interface HitCountsState {
+  hitCounts: EntityState<SourceHitCounts>;
+}
+
+export enum LoadingState {
+  LOADING = "loading",
+  LOADED = "loaded",
+  ERRORED = "errored",
+}
+
+const hitCountsAdapter = createEntityAdapter<SourceHitCounts>();
+const initialState: HitCountsState = {
+  hitCounts: hitCountsAdapter.getInitialState(),
+};
+
+const hitCountsSlice = createSlice({
+  name: "hitCounts",
+  initialState,
+  reducers: {
+    hitCountsRequested: (state, action: PayloadAction<string>) => {
+      hitCountsAdapter.upsertOne(state.hitCounts, {
+        id: action.payload,
+        status: LoadingState.LOADING,
+      });
+    },
+    hitCountsReceived: (state, action: PayloadAction<{ id: string; hitCounts: HitCount[] }>) => {
+      hitCountsAdapter.upsertOne(state.hitCounts, {
+        ...action.payload,
+        status: LoadingState.LOADED,
+      });
+    },
+    hitCountsFailed: (state, action: PayloadAction<{ id: string; error: string }>) => {
+      hitCountsAdapter.upsertOne(state.hitCounts, {
+        id: action.payload.id,
+        status: LoadingState.ERRORED,
+        error: action.payload.error,
+      });
+    },
+  },
+});
+
+const hitCountsSelectors = hitCountsAdapter.getSelectors<UIState>(
+  (state: UIState) => state.hitCounts.hitCounts
+);
+
+const MAX_LINE_HITS_TO_FETCH = 1000;
+// This will fetch hitCounts in chunks of lines. So if line 4 is request, lines
+// 1-1000 will be fetched. If line 1001 is requested, lines 1001-2000 will be
+// fetched.
+export const getBoundsForLineNumber = (line: number) => {
+  const lower = Math.floor(line / MAX_LINE_HITS_TO_FETCH) * MAX_LINE_HITS_TO_FETCH;
+  const upper = lower + MAX_LINE_HITS_TO_FETCH;
+  return {
+    lower,
+    upper,
+  };
+};
+
+export const getCacheKeyForFocusRegion = (state: UIState) => {
+  const focusRegion = getFocusRegion(state);
+  if (!focusRegion) {
+    return "";
+  }
+  const range = rangeForFocusRegion(focusRegion);
+  return `${range.begin.point}-${range.end.point}`;
+};
+
+// We need to refetch if: we are beyond the maximum number of line hits
+// fetchable OR the focusRegion has changed. So we use a cache key which will
+// change any time either of those factors change.
+export const getCacheKeyForSourceHitCounts = (
+  state: UIState,
+  sourceId: string,
+  lineNumber: number = 0
+) => {
+  const sourceDetails = getSourceDetails(state, sourceId);
+  if (!sourceDetails) {
+    throw `Source with ID ${sourceId} not found`;
+  }
+
+  const correspondingSourceIdsSection = sourceDetails.correspondingSourceIds.join("&");
+  const { lower, upper } = getBoundsForLineNumber(lineNumber);
+  const lineNumberSection = `${lower}-${upper}`;
+  const rangeSection = getCacheKeyForFocusRegion(state);
+
+  return [correspondingSourceIdsSection, lineNumberSection, rangeSection].join("|");
+};
+
+export const fetchHitCounts = (sourceId: string, lineNumber: number): UIThunkAction => {
```

Comment:
💬 this might actually be longer than the old code... but it's more _readable_ , and tbh that's what matters atm